File: apps/ios_native/Sources/PasswordManageriOSCore/Sync/VaultSyncMerger.swift
import Foundation

enum VersionComparison: Sendable {
    case equal
    case localDominates
    case remoteDominates
    case concurrent
}

enum SyncConflictStrategy: Sendable {
    case localWins
    case remoteWins
    case keepBoth
}

struct SyncMergeStats: Equatable, Sendable {
    var total: Int
    var conflicts: Int
    var deletes: Int
}

struct SyncMergeResult: Equatable, Sendable {
    var entries: [VaultEntry]
    var stats: SyncMergeStats
}

struct VaultSyncMerger {
    var idGenerator: () -> String = { UUID().uuidString.lowercased() }
    var conflictLabelBuilder: (VaultEntry, Bool) -> String = { _, _ in "(conflict)" }
    var conflictStrategy: SyncConflictStrategy = .keepBoth
    var now: () -> Date = Date.init

    func merge(localEntries: [VaultEntry], remoteEntries: [VaultEntry]) -> SyncMergeResult {
        var merged: [VaultEntry] = []
        var conflicts = 0
        var deletes = 0

        func addEntry(_ entry: VaultEntry) {
            merged.append(entry)
            if entry.isDeleted {
                deletes += 1
            }
        }

        var versionCache: [String: [String: Int]] = [:]
        func effectiveVersion(_ entry: VaultEntry) -> [String: Int] {
            if !entry.version.isEmpty {
                return entry.version
            }
            if let cached = versionCache[entry.id] {
                return cached
            }
            let updater = entry.updatedBy.isEmpty ? Self.legacyUpdater : entry.updatedBy
            let version = [updater: 1]
            versionCache[entry.id] = version
            return version
        }

        var remoteById = entriesById(remoteEntries, effectiveVersion: effectiveVersion)
        let localById = entriesById(localEntries, effectiveVersion: effectiveVersion)

        for (id, local) in localById {
            guard let remote = remoteById.removeValue(forKey: id) else {
                addEntry(local)
                continue
            }

            switch Self.compareVersion(local: effectiveVersion(local), remote: effectiveVersion(remote)) {
            case .equal:
                addEntry(pickLatest(local: local, remote: remote))
            case .localDominates:
                addEntry(local)
            case .remoteDominates:
                addEntry(remote)
            case .concurrent:
                if local.isDeleted != remote.isDeleted {
                    conflicts += 1
                    let deleted = local.isDeleted ? local : remote
                    let active = local.isDeleted ? remote : local
                    addEntry(deleted)
                    merged.append(conflictClone(active, isRemote: active.id == remote.id && active == remote))
                } else if samePayload(local: local, remote: remote) {
                    addEntry(pickLatest(local: local, remote: remote))
                } else {
                    conflicts += 1
                    let primary = choosePrimary(local: local, remote: remote)
                    let secondary = primary == local ? remote : local
                    merged.append(primary)
                    merged.append(conflictClone(secondary, isRemote: secondary == remote))
                }
            }
        }

        for remote in remoteById.values {
            addEntry(remote)
        }

        return SyncMergeResult(
            entries: merged,
            stats: SyncMergeStats(total: merged.count, conflicts: conflicts, deletes: deletes)
        )
    }

    static func compareVersion(local: [String: Int], remote: [String: Int]) -> VersionComparison {
        var localGreater = false
        var remoteGreater = false
        for key in Set(local.keys).union(remote.keys) {
            let localValue = local[key] ?? 0
            let remoteValue = remote[key] ?? 0
            if localValue > remoteValue {
                localGreater = true
            } else if remoteValue > localValue {
                remoteGreater = true
            }
            if localGreater && remoteGreater {
                return .concurrent
            }
        }
        if !localGreater && !remoteGreater {
            return .equal
        }
        return localGreater ? .localDominates : .remoteDominates
    }

    private func pickLatest(local: VaultEntry, remote: VaultEntry) -> VaultEntry {
        local.updatedAt >= remote.updatedAt ? local : remote
    }

    private func samePayload(local: VaultEntry, remote: VaultEntry) -> Bool {
        local.label == remote.label &&
            local.type == remote.type &&
            local.payload == remote.payload &&
            local.customFields == remote.customFields &&
            local.isDeleted == remote.isDeleted
    }

    private func choosePrimary(local: VaultEntry, remote: VaultEntry) -> VaultEntry {
        switch conflictStrategy {
        case .localWins:
            local
        case .remoteWins:
            remote
        case .keepBoth:
            pickLatest(local: local, remote: remote)
        }
    }

    private func entriesById(
        _ entries: [VaultEntry],
        effectiveVersion: (VaultEntry) -> [String: Int]
    ) -> [String: VaultEntry] {
        var entriesById: [String: VaultEntry] = [:]
        for entry in entries {
            guard let existing = entriesById[entry.id] else {
                entriesById[entry.id] = entry
                continue
            }
            entriesById[entry.id] = pickDuplicate(
                existing: existing,
                candidate: entry,
                effectiveVersion: effectiveVersion
            )
        }
        return entriesById
    }

    private func pickDuplicate(
        existing: VaultEntry,
        candidate: VaultEntry,
        effectiveVersion: (VaultEntry) -> [String: Int]
    ) -> VaultEntry {
        switch Self.compareVersion(local: effectiveVersion(existing), remote: effectiveVersion(candidate)) {
        case .equal:
            pickLatest(local: existing, remote: candidate)
        case .localDominates:
            existing
        case .remoteDominates:
            candidate
        case .concurrent:
            pickLatest(local: existing, remote: candidate)
        }
    }

    private func conflictClone(_ source: VaultEntry, isRemote: Bool) -> VaultEntry {
        let updatedBy = source.updatedBy.isEmpty ? Self.legacyUpdater : source.updatedBy
        let baseVersion = source.version.isEmpty ? 1 : source.version[updatedBy] ?? 1
        return VaultEntry(
            id: idGenerator(),
            label: "\(source.label) \(conflictLabelBuilder(source, isRemote))",
            type: source.type,
            payload: source.payload,
            customFields: source.customFields,
            createdAt: now(),
            updatedAt: source.updatedAt,
            isDeleted: source.isDeleted,
            deletedAt: source.deletedAt,
            version: [updatedBy: baseVersion],
            updatedBy: updatedBy
        )
    }

    private static let legacyUpdater = "legacy"
}

File: apps/android_native/app/src/main/java/life/devops/passwordmanager/sync/VaultSyncMerger.kt
package life.devops.passwordmanager.sync

import life.devops.passwordmanager.model.VaultEntry
import java.time.Instant
import java.util.Locale
import java.util.UUID

enum class VersionComparison {
    EQUAL,
    LOCAL_DOMINATES,
    REMOTE_DOMINATES,
    CONCURRENT,
}

enum class SyncConflictStrategy {
    LOCAL_WINS,
    REMOTE_WINS,
    KEEP_BOTH,
}

data class SyncMergeStats(
    val total: Int,
    val conflicts: Int,
    val deletes: Int,
)

data class SyncMergeResult(
    val entries: List<VaultEntry>,
    val stats: SyncMergeStats,
)

class VaultSyncMerger(
    private val idGenerator: () -> String = { UUID.randomUUID().toString() },
    private val conflictLabelBuilder: (VaultEntry, Boolean) -> String = { _, _ -> "(conflict)" },
    private val conflictStrategy: SyncConflictStrategy = SyncConflictStrategy.KEEP_BOTH,
    private val clock: () -> Instant = { Instant.now() },
) {
    fun merge(localEntries: List<VaultEntry>, remoteEntries: List<VaultEntry>): SyncMergeResult {
        val merged = mutableListOf<VaultEntry>()
        var conflicts = 0
        var deletes = 0

        fun addEntry(entry: VaultEntry) {
            merged += entry.withCanonicalId()
            if (entry.isDeleted) {
                deletes += 1
            }
        }

        val versionCache = mutableMapOf<String, Map<String, Int>>()
        fun effectiveVersion(entry: VaultEntry): Map<String, Int> {
            if (entry.version.isNotEmpty()) return entry.version
            return versionCache.getOrPut(entry.id) {
                mapOf((entry.updatedBy.ifBlank { LEGACY_UPDATER }) to 1)
            }
        }

        val remoteById = entriesById(remoteEntries, ::effectiveVersion).toMutableMap()
        entriesById(localEntries, ::effectiveVersion).forEach { (id, local) ->
            val remote = remoteById.remove(id)
            if (remote == null) {
                addEntry(local)
                return@forEach
            }

            when (compareVersion(effectiveVersion(local), effectiveVersion(remote))) {
                VersionComparison.EQUAL -> addEntry(pickLatest(local, remote))
                VersionComparison.LOCAL_DOMINATES -> addEntry(local)
                VersionComparison.REMOTE_DOMINATES -> addEntry(remote)
                VersionComparison.CONCURRENT -> {
                    if (local.isDeleted != remote.isDeleted) {
                        conflicts += 1
                        val deleted = if (local.isDeleted) local else remote
                        val active = if (local.isDeleted) remote else local
                        addEntry(deleted)
                        merged += conflictClone(active, isRemote = active === remote)
                    } else if (samePayload(local, remote)) {
                        addEntry(pickLatest(local, remote))
                    } else {
                        conflicts += 1
                        val primary = choosePrimary(local, remote)
                        val secondary = if (primary === local) remote else local
                        merged += primary
                        merged += conflictClone(secondary, isRemote = secondary === remote)
                    }
                }
            }
        }

        remoteById.values.forEach(::addEntry)

        return SyncMergeResult(
            entries = merged,
            stats = SyncMergeStats(
                total = merged.size,
                conflicts = conflicts,
                deletes = deletes,
            ),
        )
    }

    private fun pickLatest(local: VaultEntry, remote: VaultEntry): VaultEntry =
        if (local.updatedAt >= remote.updatedAt) local else remote

    private fun samePayload(local: VaultEntry, remote: VaultEntry): Boolean =
        local.label == remote.label &&
            local.type == remote.type &&
            local.payload == remote.payload &&
            local.customFields == remote.customFields &&
            local.isDeleted == remote.isDeleted

    private fun choosePrimary(local: VaultEntry, remote: VaultEntry): VaultEntry =
        when (conflictStrategy) {
            SyncConflictStrategy.LOCAL_WINS -> local
            SyncConflictStrategy.REMOTE_WINS -> remote
            SyncConflictStrategy.KEEP_BOTH -> pickLatest(local, remote)
        }

    private fun entriesById(
        entries: List<VaultEntry>,
        effectiveVersion: (VaultEntry) -> Map<String, Int>,
    ): Map<String, VaultEntry> {
        val entriesById = mutableMapOf<String, VaultEntry>()
        entries.forEach { entry ->
            val canonicalId = entry.id.canonicalUuidString()
            val canonicalEntry = entry.withCanonicalId(canonicalId)
            val existing = entriesById[canonicalId]
            entriesById[canonicalId] = if (existing == null) {
                canonicalEntry
            } else {
                pickDuplicate(existing, canonicalEntry, effectiveVersion)
            }
        }
        return entriesById
    }

    private fun pickDuplicate(
        existing: VaultEntry,
        candidate: VaultEntry,
        effectiveVersion: (VaultEntry) -> Map<String, Int>,
    ): VaultEntry =
        when (compareVersion(effectiveVersion(existing), effectiveVersion(candidate))) {
            VersionComparison.EQUAL -> pickLatest(existing, candidate)
            VersionComparison.LOCAL_DOMINATES -> existing
            VersionComparison.REMOTE_DOMINATES -> candidate
            VersionComparison.CONCURRENT -> pickLatest(existing, candidate)
        }

    private fun conflictClone(source: VaultEntry, isRemote: Boolean): VaultEntry {
        val updatedBy = source.updatedBy.ifBlank { LEGACY_UPDATER }
        val baseVersion = if (source.version.isNotEmpty()) {
            source.version[updatedBy] ?: 1
        } else {
            1
        }
        return source.copy(
            id = idGenerator(),
            label = "${source.label} ${conflictLabelBuilder(source, isRemote)}",
            createdAt = clock(),
            version = mapOf(updatedBy to baseVersion),
            updatedBy = updatedBy,
        )
    }

    companion object {
        private const val LEGACY_UPDATER = "legacy"

        fun compareVersion(local: Map<String, Int>, remote: Map<String, Int>): VersionComparison {
            var localGreater = false
            var remoteGreater = false
            (local.keys + remote.keys).forEach { key ->
                val localValue = local[key] ?: 0
                val remoteValue = remote[key] ?: 0
                if (localValue > remoteValue) {
                    localGreater = true
                } else if (remoteValue > localValue) {
                    remoteGreater = true
                }
                if (localGreater && remoteGreater) {
                    return VersionComparison.CONCURRENT
                }
            }
            return when {
                !localGreater && !remoteGreater -> VersionComparison.EQUAL
                localGreater -> VersionComparison.LOCAL_DOMINATES
                remoteGreater -> VersionComparison.REMOTE_DOMINATES
                else -> VersionComparison.CONCURRENT
            }
        }
    }
}

private fun VaultEntry.withCanonicalId(id: String = this.id.canonicalUuidString()): VaultEntry =
    if (this.id == id) this else copy(id = id)

private fun String.canonicalUuidString(): String {
    val trimmed = trim()
    if (trimmed.isEmpty()) return trimmed
    return runCatching { UUID.fromString(trimmed).toString() }
        .getOrElse { trimmed.lowercase(Locale.ROOT) }
}
